### PR TITLE
try to enhance conda build compatibility

### PR DESCRIPTION
**Pull request by jaimergp**
Support PR for https://github.com/conda/conda-build/pull/4431

### CODE REVIEW

**Comment by jaimergp** [comment]
All tests in conda-build now passing! 🥳 

I am tempted to add upstream's CI here as well, but right now `conda-build`'s CI is not really ready for copy-pasting (broken in different ways). I'll argue we need to do that in the future, but for now we'll wait til upstream is operational. 

Additionally, conda-build's usage of the solver is more straight forward (basically new envs, that's it), so there's way less surface for bugs. Potential users are expected to be more savvy and less numerous too.

In short, yes to upstream CI, but when it's in better shape because although important, this is not urgent right now.

**Comment by jezdez** [comment]
> All tests in conda-build now passing! 🥳
> 
> I am tempted to add upstream's CI here as well, but right now `conda-build`'s CI is not really ready for copy-pasting (broken in different ways). I'll argue we need to do that in the future, but for now we'll wait til upstream is operational.
> 
> Additionally, conda-build's usage of the solver is more straight forward (basically new envs, that's it), so there's way less surface for bugs. Potential users are expected to be more savvy and less numerous too.
> 
> In short, yes to upstream CI, but when it's in better shape because although important, this is not urgent right now.

A few hours after your comment https://github.com/conda/conda-build/commit/5e5707bf99b311e5e43fba449dffdee44265db02 got merged which should make things a bit easier to port over. WDYT?

**Comment by jaimergp** [comment]
> A few hours after your comment https://github.com/conda/conda-build/commit/5e5707bf99b311e5e43fba449dffdee44265db02 got merged which should make things a bit easier to port over. WDYT?

That's awesome! I would still wait to bring parts of that CI over, because maybe we find some stuff not working perfectly over the next couple weeks. Let's revisit in June?

**Comment by jaimergp** [comment]
@jezdez - I have added some tests (generously borrowed from boa/mambabuild :D) to ensure we can test _some_ conda-build jobs in our CI without vendoring the full thing. Removing the changes in this PR breaks the tests successfully, so I'd say we are good to merge if you approve the submission!

File: conda_libmamba_solver/solver.py [hunk 1]
```diff
@@ -797,3 +881,12 @@ def _check_spec_compat(self, match_spec):
...
         self._solver_options = None
+
+    def _called_from_conda_build(self):
+        """
+        conda build calls the solver via `conda.plan.install_actions`, which
+        overrides Solver._index (populated in the classic solver, but empty for us)
+        with a custom index. We can use this to detect whether conda build is in use
+        and apply some compatibility fixes.
+        """
+        return "conda_build" in sys.modules and getattr(self, "_index", None)
```
Comment (jezdez) [inline conda_libmamba_solver/solver.py:892]: This might make sense to use the stdlib [`inspect.currentframe()`](https://docs.python.org/3/library/inspect.html#inspect.currentframe) here instead of checking for `sys.modules` since the latter might be true for other weird reasons and could create a false positive.

E.g.:
```suggestion
        return "install_actions" in inspect.currentframe().f_back.f_code.co_name and getattr(self, "_index", None)
```

**Review by jezdez — APPROVED** [review]
This isn't super exciting since it just shows that conda-build needs a proper API to do this type of integration, but in the meantime...

Let's use inspect instead of the `sys.modules` hack.

Also, could you please also file a ticket in conda-build to allow a non-hacky implementation to occur?

**Review by jezdez — APPROVED** [review]

File: conda_libmamba_solver/solver.py [hunk 2]
```diff
@@ -641,12 +672,39 @@ def _specs_to_tasks_remove(self, in_state: SolverInputState, out_state: SolverOu
...
+            if spec.version:
+                version_str = str(spec.version)
+                if re.match(only_dot_or_digit_re, version_str):
+                    spec_fields = spec.conda_build_form().split()
+                    if version_str.count(".") <= 1:
+                        spec_fields[1] = version_str + ".*"
+                    else:
+                        spec_fields[1] = version_str + "*"
+                    return MatchSpec(" ".join(spec_fields))
+            return spec
```
Comment (jezdez) [inline conda_libmamba_solver/solver.py:693]: Let's move this out of the function, to reduce the effort needed to define inline functions.

**Review by jezdez — APPROVED** [review]
Just a new nits, nothing major. nicely done!

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -54,6 +54,7 @@ jobs:
           ghcr.io/conda/conda-ci:master-linux-python${{ matrix.python-version }}
           bash -c "sudo env CONDA_EXPERIMENTAL_SOLVER=classic \
                         /opt/conda/condabin/conda install -p /opt/conda \
+                        --repodata-fn=repodata.json --update-specs \
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply .github/workflows/ci.yml:57]: It doesn't hurt. It should alleviate some pressure on the solver so it doesn't try to freeze the installed packages (and it can fail faster too).

File: .github/workflows/upstream_ci.yml [hunk 4]
```diff
@@ -236,7 +236,7 @@ jobs:
           if errorlevel 1 exit 1
           call .\dev-init.bat
           if errorlevel 1 exit 1
-          python -c "from mamba import __version__; print('mamba', __version__)"
+            python -c "from importlib_metadata import version; print('libmambapy', version('libmambapy'))"
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply .github/workflows/upstream_ci.yml:239]: Good catch :)

File: .github/workflows/ci.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jezdez) [reply .github/workflows/ci.yml:57]: Sounds good.
